File: truevfs-kernel/truevfs-kernel-impl/src/main/scala/net/java/truevfs/kernel/impl/ArchiveController.scala
/*
 * Copyright (C) 2005-2012 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl

import net.java.truecommons.shed._
import java.io._
import java.nio.file._
import net.java.truevfs.kernel.spec._
import net.java.truecommons.cio._
import net.java.truecommons.cio.Entry._

/** Provides read/write access to an archive file system.
  * This is a mirror of [[net.java.truevfs.kernel.spec.FsController]] which has
  * been customozed to leverage the Scala language and library.
  *
  * @author Christian Schlichtherle
  */
private trait ArchiveController[E <: FsArchiveEntry] {

  /** Returns the archive model.
    *
    * @return The archive model.
    */
  def model: ArchiveModel[E]

  def node(options: AccessOptions, name: FsNodeName): Option[FsNode]
  def checkAccess(options: AccessOptions, name: FsNodeName, types: BitField[Access])
  def setReadOnly(options: AccessOptions, name: FsNodeName)
  def setTime(options: AccessOptions, name: FsNodeName, times: Map[Access, Long]): Boolean
  def setTime(options: AccessOptions, name: FsNodeName, types: BitField[Access], value: Long): Boolean
  def input(options: AccessOptions, name: FsNodeName): AnyInputSocket
  def output(options: AccessOptions, name: FsNodeName, template: Option[Entry]): AnyOutputSocket
  def make(options: AccessOptions, name: FsNodeName, tµpe: Type, template: Option[Entry])
  def unlink(options: AccessOptions, name: FsNodeName)
  def sync(options: SyncOptions)

  /** Two archive controllers are considered equal if and only if they are
    * identical.
    *
    * @param  that the object to compare.
    * @return `this == that`
    */
  final override def equals(that: Any) = this == that

  /** Returns a hash code which is consistent with `equals`.
    *
    * @return A hash code which is consistent with `equals`.
    * @see    Object#hashCode
    */
   final override def hashCode = System.identityHashCode(this)

  /** Returns a string representation of this object for logging and debugging
    * purposes.
    */
  final override def toString =
    "%s@%x[model=%s]".format(getClass.getName, hashCode, model)
}

File: truevfs-kernel/truevfs-kernel-impl/src/main/java/net/java/truevfs/kernel/impl/ArchiveController.java
/*
 * Copyright (C) 2005-2020 Schlichtherle IT Services.
 * All rights reserved. Use is subject to license terms.
 */
package net.java.truevfs.kernel.impl;

import net.java.truecommons.cio.Entry;
import net.java.truecommons.cio.Entry.Access;
import net.java.truecommons.cio.Entry.Type;
import net.java.truecommons.cio.InputSocket;
import net.java.truecommons.cio.OutputSocket;
import net.java.truecommons.shed.BitField;
import net.java.truevfs.kernel.spec.*;

import java.io.IOException;
import java.util.Map;
import java.util.Optional;

/**
 * Provides read/write access to an archive file system.
 * This is a mirror of {@link net.java.truevfs.kernel.spec.FsController} which has been customized to ease the
 * implementation.
 *
 * @author Christian Schlichtherle
 */
interface ArchiveController<E extends FsArchiveEntry> extends ArchiveModelAspect<E>, ReentrantReadWriteLockAspect {

    Optional<? extends FsNode> node(BitField<FsAccessOption> options, FsNodeName name) throws IOException;

    void checkAccess(BitField<FsAccessOption> options, FsNodeName name, BitField<Access> types) throws IOException;

    void setReadOnly(BitField<FsAccessOption> options, FsNodeName name) throws IOException;

    boolean setTime(BitField<FsAccessOption> options, FsNodeName name, Map<Access, Long> times) throws IOException;

    boolean setTime(BitField<FsAccessOption> options, FsNodeName name, BitField<Access> types, long time) throws IOException;

    InputSocket<? extends Entry> input(BitField<FsAccessOption> options, FsNodeName name);

    OutputSocket<? extends Entry> output(BitField<FsAccessOption> options, FsNodeName name, Optional<Entry> template);

    void make(BitField<FsAccessOption> options, FsNodeName name, Type type, Optional<Entry> template) throws IOException;

    void unlink(BitField<FsAccessOption> options, FsNodeName name) throws IOException;

    void sync(BitField<FsSyncOption> options) throws FsSyncException;
}
